File: cloudflare/src/github.py
"""
GitHub data fetching layer for the Cloudflare Worker.

Uses JS `fetch` and `Headers` from Pyodide to make HTTP requests,
as the Cloudflare Workers Python runtime does not support `requests`.
"""

from js import fetch, Headers
import json
import urllib.parse


def _make_headers(d):
    """Convert a Python dict of headers into a JS Headers object.

    Input: d — a dict of {header-name: value} strings.
    Return: a JS Headers object with each pair set on it.

    CF Workers Python (Pyodide) requires JS Headers objects, not Python
    dicts, for the fetch() `headers` option, hence this adapter.
    """
    h = Headers.new()
    for k, v in d.items():
        h.set(k, v)
    return h


_UA = {"User-Agent": "VyOS-Config-Worker"}
_DOH_PROVIDERS = [
    (
        "google-json",
        "https://dns.google/resolve?name={qname}&type=A",
    ),
    (
        "cloudflare-json",
        "https://cloudflare-dns.com/dns-query?name={qname}&type=A",
    ),
]


async def github_raw(user, repo, path):
    """Fetch one file's raw text from GitHub.

    Input: user, repo, path — locate the file at
    raw.githubusercontent.com/{user}/{repo}/main/{path} (always the `main`
    branch).
    Return: the file body as a str, or None if the response status is not 200
    (e.g. 404 missing file). Network/fetch errors propagate as exceptions.
    """
    url = f"https://raw.githubusercontent.com/{user}/{repo}/main/{path}"
    resp = await fetch(url, {"headers": _make_headers(_UA)})
    if resp.status != 200:
        return None
    return await resp.text()


async def github_raw_json(user, repo, path):
    """Fetch one file from GitHub and parse it as JSON.

    Input: user, repo, path — same locator as github_raw().
    Return: the parsed JSON value (dict/list/...), or None if the file is
    missing (status != 200). Raises json.JSONDecodeError if the body is not
    valid JSON.
    """
    text = await github_raw(user, repo, path)
    if text is None:
        return None
    return json.loads(text)


async def load_yaml_config(user, config_repo):
    """Load and parse the router config YAML from the config repo.

    Input: user, config_repo — the file is fetched from
    {config_repo}/network/vyos/vyos.yaml on `main`.
    Return: the parsed YAML as a Python dict, or None if the file is missing.
    Raises yaml.YAMLError if the body is not valid YAML.
    """
    import yaml

    text = await github_raw(user, config_repo, "network/vyos/vyos.yaml")
    if text is None:
        return None
    return yaml.safe_load(text)


def _extract_a_records(data):
    """Pull the unique IPv4 answers out of a DoH JSON response.

    Input: data — a parsed DNS-over-HTTPS JSON response (dict with an
    "Answer" list).
    Return: a sorted list of unique IPv4 address strings, taken from answers
    whose type == 1 (A record). Empty list if there are no A answers.
    """
    return sorted(
        {
            answer["data"]
            for answer in data.get("Answer", [])
            if answer.get("type") == 1 and "data" in answer
        }
    )


def _shorten_error_body(text, limit=120):
    """Collapse and truncate an upstream error body for log lines.

    Input: text — an arbitrary response body; limit — max output length.
    Return: the body with all runs of whitespace collapsed to single spaces,
    truncated to `limit` chars (last 3 replaced with "..." when it overflows)
    so it fits on one worker-trace line.
    """
    compact = " ".join(text.split())
    if len(compact) <= limit:
        return compact
    return compact[: limit - 3] + "..."


async def resolve_router_id(router_name):
    """Resolve a router hostname to its single IPv4 address via DNS-over-HTTPS.

    Input: router_name — the hostname to resolve (used as the router-id).
    Return: the resolved IPv4 address as a str, returned as soon as a provider
    yields exactly one A record.

    Tries each provider in _DOH_PROVIDERS in order, accumulating a per-provider
    error reason whenever one fails (fetch/transport error, non-200, invalid
    JSON, DNS error status, or no A record), and failing over to the next.
    Raises ValueError if the name resolves to more than one address (ambiguous
    router-id) or if every provider fails (the message lists all collected
    reasons).
    """
    qname = urllib.parse.quote(router_name, safe="")
    errors = []

    for provider_name, url_template in _DOH_PROVIDERS:
        url = url_template.format(qname=qname)
        try:
            resp = await fetch(
                url,
                {
                    "method": "GET",
                    "headers": _make_headers(
                        {**_UA, "Accept": "application/dns-json"}
                    ),
                },
            )
            text = await resp.text()
        except Exception as e:
            # Transport-level failure (unreachable provider, TLS error,
            # timeout): treat like any other provider failure and fail over
            # to the next one rather than aborting the whole resolution.
            errors.append(f"{provider_name}=fetch error ({e})")
            continue
        if resp.status != 200:
            errors.append(
                f"{provider_name}=HTTP {resp.status} ({_shorten_error_body(text)})"
            )
            continue

        try:
            data = json.loads(text)
        except json.JSONDecodeError:
            errors.append(
                f"{provider_name}=invalid JSON ({_shorten_error_body(text)})"
            )
            continue

        answers = _extract_a_records(data)
        if len(answers) == 1:
            return answers[0]
        if len(answers) > 1:
            raise ValueError(
                f"router-id for {router_name} is not unique: {', '.join(answers)}"
            )

        status = data.get("Status")
        if status not in (None, 0):
            errors.append(f"{provider_name}=DNS status {status}")
        else:
            errors.append(f"{provider_name}=no A record")

    raise ValueError(
        f"DNS query failed for {router_name}: {'; '.join(errors)}"
    )

File: cloudflare/src/github.ts
/**
 * GitHub data fetching layer (port of github.py).
 *
 * Uses the native Workers fetch/Headers/AbortSignal (no Pyodide shim).
 */

import yaml from "js-yaml";

const UA = { "User-Agent": "VyOS-Config-Worker" };

// Subrequest timeouts: a stalled origin otherwise hangs the whole request.
const HTTP_TIMEOUT_MS = 3000;
const DOH_TIMEOUT_MS = 3000;

const DOH_PROVIDERS: [string, string][] = [
  ["google-json", "https://dns.google/resolve?name={qname}&type=A"],
  ["cloudflare-json", "https://cloudflare-dns.com/dns-query?name={qname}&type=A"],
];

/** Percent-encode like Python urllib.parse.quote(safe=""). */
function quote(s: string): string {
  return encodeURIComponent(s).replace(
    /[!*'()]/g,
    (c) => "%" + c.charCodeAt(0).toString(16).toUpperCase(),
  );
}

/** Fetch one file's raw text from GitHub `main`; null on non-200. */
export async function githubRaw(
  user: string,
  repo: string,
  path: string,
): Promise<string | null> {
  // Always read fresh from GitHub. Two caches sit between us and the file:
  //   1. Cloudflare's subrequest cache. A Worker fetch() caches by URL, honoring
  //      GitHub raw's `Cache-Control: max-age=300` — this is what served stale
  //      configs. `cache: "no-store"` makes the runtime skip that cache entirely
  //      (for origins not on Cloudflare it bypasses Cloudflare's cache and also
  //      forwards `Cache-Control: no-cache` upstream). Supported since compat
  //      date 2024-11-11; ours is 2024-12-01, so no extra flag is needed.
  //   2. GitHub raw's own CDN (Fastly). The unique query string is a best-effort
  //      bust for that layer, varying the origin cache key when it's honored.
  const url =
    `https://raw.githubusercontent.com/${user}/${repo}/main/${path}` +
    `?nocache=${Date.now()}`;
  const resp = await fetch(url, {
    // Honored at runtime (compat date 2024-12-01 ≥ 2024-11-11); the intersection
    // cast is only because this @cloudflare/workers-types version omits `cache`.
    cache: "no-store",
    headers: new Headers(UA),
    signal: AbortSignal.timeout(HTTP_TIMEOUT_MS),
  } as RequestInit & { cache: "no-store" });
  if (resp.status !== 200) return null;
  return await resp.text();
}

/** Fetch one file from GitHub and parse it as JSON; null if missing. */
export async function githubRawJson(
  user: string,
  repo: string,
  path: string,
): Promise<any> {
  const text = await githubRaw(user, repo, path);
  if (text === null) return null;
  return JSON.parse(text);
}

/** Load and parse the router config YAML from the config repo; null if missing. */
export async function loadYamlConfig(
  user: string,
  configRepo: string,
): Promise<any> {
  const text = await githubRaw(user, configRepo, "network/vyos/vyos.yaml");
  if (text === null) return null;
  return yaml.load(text);
}

function extractARecords(data: any): string[] {
  const set = new Set<string>();
  for (const answer of data?.Answer ?? []) {
    if (answer?.type === 1 && "data" in answer) set.add(answer.data);
  }
  return [...set].sort();
}

function shortenErrorBody(text: string, limit = 120): string {
  const compact = text.split(/\s+/).filter(Boolean).join(" ");
  if (compact.length <= limit) return compact;
  return compact.slice(0, limit - 3) + "...";
}

/**
 * Resolve a router hostname to its single IPv4 via DNS-over-HTTPS, failing over
 * across providers. Throws if the name is ambiguous or every provider fails.
 */
export async function resolveRouterId(routerName: string): Promise<string> {
  const qname = quote(routerName);
  const errors: string[] = [];

  for (const [providerName, urlTemplate] of DOH_PROVIDERS) {
    const url = urlTemplate.replace("{qname}", qname);
    let text: string;
    let resp: Response;
    try {
      resp = await fetch(url, {
        method: "GET",
        headers: new Headers({ ...UA, Accept: "application/dns-json" }),
        signal: AbortSignal.timeout(DOH_TIMEOUT_MS),
      });
      text = await resp.text();
    } catch (e) {
      errors.push(`${providerName}=fetch error (${e})`);
      continue;
    }
    if (resp.status !== 200) {
      errors.push(`${providerName}=HTTP ${resp.status} (${shortenErrorBody(text)})`);
      continue;
    }

    let data: any;
    try {
      data = JSON.parse(text);
    } catch {
      errors.push(`${providerName}=invalid JSON (${shortenErrorBody(text)})`);
      continue;
    }

    const answers = extractARecords(data);
    if (answers.length === 1) return answers[0];
    if (answers.length > 1) {
      throw new Error(
        `router-id for ${routerName} is not unique: ${answers.join(", ")}`,
      );
    }

    const status = data?.Status;
    if (status !== undefined && status !== null && status !== 0) {
      errors.push(`${providerName}=DNS status ${status}`);
    } else {
      errors.push(`${providerName}=no A record`);
    }
  }

  throw new Error(`DNS query failed for ${routerName}: ${errors.join("; ")}`);
}
